use-cases/extras/solutions.md

```diff
@@ -0,0 +1,110 @@
+# Solutions Exploration
+The purpose of this document is to explore possible hooks and/or APIs that may be used to resolve at least some of the use-cases presented in the "extras" section of this repo.
+
+Suggestions will be listed and discussed in an arbitrary order and may contain overlaps in functionality as well as in the problems they intend to solve.
+
+## Suggestion I - Async Functions Controller
+**Idea:** Give the invoker of an async function (some) ability to control its execution in a backward compatible way.
+
+### Example
+The following code should print out `Waited 50 ms`:
+```js
+// Does not exist:
+const {getController} = require('async_hooks');
+
+const setTimeoutPromise = require('util').promisify(setTimeout);
+
+async function wait(ms) {
+  const before = Date.now();
+
+  try {
+    await setTimeoutPromise(ms);
+  }
+  finally {
+    console.log("Waited", Date.now() - before, 'ms');
+  }
+}
+
+let p = wait(100);
+const controller = getController(p);
+
+setTimeoutPromise(50)
+  .then(() => controller.return());
+```
+Since the `getController` API only works on promises returned from async function invocations, the invoker can have full control over who gets to abort the operation.
```

Comment:
How do you differentiate between Promises created by async functions and ones created with the new Promise constructor? Do you have an internal flag or something?

Reply:
It would really be up to the implementer, but I imagine you can internally map promises created for async functions to their controllers. Other kinds of promises would simply have no mapping (and no controller).